# Exercicis per practicar amb regressions.

In [1]:
import pandas as pd

In [2]:
print(open('housing data description.txt').read())

1. Title: Boston Housing Data

2. Sources:
   (a) Origin:  This dataset was taken from the StatLib library which is
                maintained at Carnegie Mellon University.
   (b) Creator:  Harrison, D. and Rubinfeld, D.L. 'Hedonic prices and the 
                 demand for clean air', J. Environ. Economics & Management,
                 vol.5, 81-102, 1978.
   (c) Date: July 7, 1993

3. Past Usage:
   -   Used in Belsley, Kuh & Welsch, 'Regression diagnostics ...', Wiley, 
       1980.   N.B. Various transformations are used in the table on
       pages 244-261.
    -  Quinlan,R. (1993). Combining Instance-Based and Model-Based Learning.
       In Proceedings on the Tenth International Conference of Machine 
       Learning, 236-243, University of Massachusetts, Amherst. Morgan
       Kaufmann.

4. Relevant Information:

   Concerns housing values in suburbs of Boston.

5. Number of Instances: 506

6. Number of Attributes: 13 continuous attributes (including "class"
                

Arriba tenemos una descripción teórica de los datos. Estudiamos diversas variables que afectan al precio de las viviendas en los suburbios de Boston en los años 70. `MEDV` es la variable dependiente que refleja los precios, mientras el resto de características tales como la tasa de crímenes, si está junto al río o no, el nivel de accesibilidad a las autopistas, % de población de clase baja, distancia a los centros de empleo de Boston, precio mediano de las viviendas ... las trataremos como variables independientes.

In [3]:
columnas = ['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD', 'TAX', 'PTRATIO', 'B', 'LSTAT', 'MEDV']
viviendas = pd.read_csv('housing data.csv', sep=',', names=columnas)
viviendas.head(2)

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296.0   
1  0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242.0   

   PTRATIO      B  LSTAT  MEDV  
0     15.3  396.9   4.98  24.0  
1     17.8  396.9   9.14  21.6

Lo primero que hemos hecho ha sido cargar todos los datos del dataframe `housing data.csv` en la variable viviendas y asignar los nombres de las columnas del mismo, como podemos ver en la muestra de las dos primeras filas.

A continuación podemos ver las distintas variables que contiene la base de datos:

In [4]:
viviendas.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 14 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   CRIM     506 non-null    float64
 1   ZN       506 non-null    float64
 2   INDUS    506 non-null    float64
 3   CHAS     506 non-null    int64  
 4   NOX      506 non-null    float64
 5   RM       506 non-null    float64
 6   AGE      506 non-null    float64
 7   DIS      506 non-null    float64
 8   RAD      506 non-null    int64  
 9   TAX      506 non-null    float64
 10  PTRATIO  506 non-null    float64
 11  B        506 non-null    float64
 12  LSTAT    506 non-null    float64
 13  MEDV     506 non-null    float64
dtypes: float64(12), int64(2)
memory usage: 55.5 KB


Como podemos comprobar no existen registros nulos y todos los datos son de tipo numérico, para un total de 506 registros. Es interesante también observar como tanto `CHAS` como `RAD` son de tipo entero (int), lo que nos da una idea de su carácter categórico, mientras el resto de variables son de coma flotante.

In [5]:
viviendas.corr().round(2)

CRIM    ZN  INDUS  CHAS   NOX    RM   AGE   DIS   RAD   TAX  PTRATIO  \
CRIM     1.00 -0.20   0.41 -0.06  0.42 -0.22  0.35 -0.38  0.63  0.58     0.29   
ZN      -0.20  1.00  -0.53 -0.04 -0.52  0.31 -0.57  0.66 -0.31 -0.31    -0.39   
INDUS    0.41 -0.53   1.00  0.06  0.76 -0.39  0.64 -0.71  0.60  0.72     0.38   
CHAS    -0.06 -0.04   0.06  1.00  0.09  0.09  0.09 -0.10 -0.01 -0.04    -0.12   
NOX      0.42 -0.52   0.76  0.09  1.00 -0.30  0.73 -0.77  0.61  0.67     0.19   
RM      -0.22  0.31  -0.39  0.09 -0.30  1.00 -0.24  0.21 -0.21 -0.29    -0.36   
AGE      0.35 -0.57   0.64  0.09  0.73 -0.24  1.00 -0.75  0.46  0.51     0.26   
DIS     -0.38  0.66  -0.71 -0.10 -0.77  0.21 -0.75  1.00 -0.49 -0.53    -0.23   
RAD      0.63 -0.31   0.60 -0.01  0.61 -0.21  0.46 -0.49  1.00  0.91     0.46   
TAX      0.58 -0.31   0.72 -0.04  0.67 -0.29  0.51 -0.53  0.91  1.00     0.46   
PTRATIO  0.29 -0.39   0.38 -0.12  0.19 -0.36  0.26 -0.23  0.46  0.46     1.00   
B       -0.39  0.18  -0.36  0.05 -0.38  0.13 -0.27  0.29 -0.44 -0.44    -0.18   
LSTAT    0.46 -0.41   0.60 -0.05  0.59 -0.61  0.60 -0.50  0.49  0.54     0.37   
MEDV    -0.39  0.36  -0.48  0.18 -0.43  0.70 -0.38  0.25 -0.38 -0.47    -0.51   

            B  LSTAT  MEDV  
CRIM    -0.39   0.46 -0.39  
ZN       0.18  -0.41  0.36  
INDUS   -0.36   0.60 -0.48  
CHAS     0.05  -0.05  0.18  
NOX     -0.38   0.59 -0.43  
RM       0.13  -0.61  0.70  
AGE     -0.27   0.60 -0.38  
DIS      0.29  -0.50  0.25  
RAD     -0.44   0.49 -0.38  
TAX     -0.44   0.54 -0.47  
PTRATIO -0.18   0.37 -0.51  
B        1.00  -0.37  0.33  
LSTAT   -0.37   1.00 -0.74  
MEDV     0.33  -0.74  1.00

Aquí podemos ver como se distribuyen las distintas correlaciones entre las variables sin transformar, en general son correlaciones medias, tanto positivas como negativas, dependiendo de las variables. `CHAS` al ser binaria está naturalmente menos correlacionada con todas las demás.

In [6]:
viviendas_num = viviendas.drop(['CHAS','RAD'], axis=1)
viviendas_num.describe().round(2)

CRIM      ZN   INDUS     NOX      RM     AGE     DIS     TAX  \
count  506.00  506.00  506.00  506.00  506.00  506.00  506.00  506.00   
mean     3.61   11.36   11.14    0.55    6.28   68.57    3.80  408.24   
std      8.60   23.32    6.86    0.12    0.70   28.15    2.11  168.54   
min      0.01    0.00    0.46    0.38    3.56    2.90    1.13  187.00   
25%      0.08    0.00    5.19    0.45    5.89   45.02    2.10  279.00   
50%      0.26    0.00    9.69    0.54    6.21   77.50    3.21  330.00   
75%      3.68   12.50   18.10    0.62    6.62   94.07    5.19  666.00   
max     88.98  100.00   27.74    0.87    8.78  100.00   12.13  711.00   

       PTRATIO       B   LSTAT    MEDV  
count   506.00  506.00  506.00  506.00  
mean     18.46  356.67   12.65   22.53  
std       2.16   91.29    7.14    9.20  
min      12.60    0.32    1.73    5.00  
25%      17.40  375.38    6.95   17.02  
50%      19.05  391.44   11.36   21.20  
75%      20.20  396.22   16.96   25.00  
max      22.00  396.90   37.97   50.00

Arriba mostramos los principales parámetros de centralización y dispersión dentro de cada una de las variables, excluyendo aquellas que tienen un carácter binario / categórico (`CHAS` y `RAD`). Se observa como las escalas son distintas entre ellas, las medias están alejadas de 0, las desviciones típicas y rango intercuartílico (75% - 25%) reflejan a los datos alejados de la media y probablemente también existan valores extremos.

Por todo lo anterior sería adecuado hacer un preprocesamiento de los datos que nos permita ajustar mejor nuestros modelos a la muestra disponible.

# Preprocesamiento de datos:

En primer lugar, vamos a separar la variable dependiente y, de las variables dependientes X.

In [7]:
y = viviendas.MEDV
X = viviendas.drop('MEDV', axis=1)

En segundo lugar, separaremos los conjuntos de entrenamiento y prueba, antes de hacer cualquier transformación, para evitar sesgar los datos de prueba al transformarlos con los datos de entrenamiento, lo que haría menos objetivo el contraste del modelo.

In [8]:
from sklearn.model_selection import train_test_split

In [9]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=1)

Una vez hecho lo anterior, ya podemos transformar nuestras variables, lo que uniformiza los datos y en definitiva nos permite obtener modelos más precisos.

Vamos primero a comprobar cuáles de nuestras variables siguen una distribución normal:

In [10]:
from scipy.stats import shapiro

In [11]:
for x in X_train:
  stat, p = shapiro(X_train[x])
  if p > 0.05:
    print(x+': Gaussiana'+' - p='+str(round(p,3)))
  else:
    print(x+': No gaussiana'+' - p='+str(round(p,3)))

CRIM: No gaussiana - p=0.0
ZN: No gaussiana - p=0.0
INDUS: No gaussiana - p=0.0
CHAS: No gaussiana - p=0.0
NOX: No gaussiana - p=0.0
RM: No gaussiana - p=0.0
AGE: No gaussiana - p=0.0
DIS: No gaussiana - p=0.0
RAD: No gaussiana - p=0.0
TAX: No gaussiana - p=0.0
PTRATIO: No gaussiana - p=0.0
B: No gaussiana - p=0.0
LSTAT: No gaussiana - p=0.0


De acuerdo con el test de Shapiro-Wilk ninguna de las variables del conjunto de datos de entrenamiento sigue una distribución normal (gaussiana).

Recuperamos las transformaciones hechas en la práctica M5_T02:

Numéricas no gaussianas sin valores extremos: MinMax: INDUS, NOX, AGE y TAX.

In [22]:
X_train_transform = pd.DataFrame()
X_test_transform = pd.DataFrame()

In [23]:
from sklearn.preprocessing import MinMaxScaler

scaler = MinMaxScaler()
min_max_cols = ['INDUS', 'NOX', 'AGE', 'TAX']

for i in min_max_cols:
  X_train_min_max = scaler.fit_transform(X_train[[i]]).reshape(-1)
  X_train_min_max_pd = pd.Series(X_train_min_max)
  X_train_transform[i] = X_train_min_max_pd
  X_test_min_max = scaler.transform(X_test[[i]]).reshape(-1)
  X_test_min_max_pd = pd.Series(X_test_min_max)
  X_test_transform[i] = X_test_min_max_pd

Ahora ya tenemos todas las variables no gaussianas sin valores extremos transformadas y guardadas en viviendas_train_transform y viviendas_test_transform. Sigamos ...

In [24]:
from sklearn.preprocessing import RobustScaler

scaler = RobustScaler()
robust_cols = ['CRIM', 'ZN', 'RM', 'DIS', 'PTRATIO', 'B', 'LSTAT']

for i in robust_cols:
  X_train_robust = scaler.fit_transform(X_train[[i]]).reshape(-1)
  X_train_robust_pd = pd.Series(X_train_robust)
  X_train_transform[i] = X_train_robust_pd
  X_test_robust = scaler.transform(X_test[[i]]).reshape(-1)
  X_test_robust_pd = pd.Series(X_test_robust)
  X_test_transform[i] = X_test_robust_pd

... y ahora con valores extremos, utilizando RobustScaler.

Nos quedan las variables de carácter binario / categórico: `CHAS` y `RAD`.

Para `CHAS` lo único que tenemos que hacer es incorporarla al nuevo dataframe creado para guardar las variables transformadas:

In [25]:
X_train_transform['CHAS'] = X_train['CHAS'].values
X_test_transform['CHAS'] = X_test['CHAS'].values

Mientras que `RAD` la desglosaremos en 9 columnas de tipo dummy:

In [27]:
from sklearn.preprocessing import OneHotEncoder

scaler = OneHotEncoder()

X_train_RAD = scaler.fit_transform(X_train[['RAD']])
X_train_RAD_df = pd.DataFrame(X_train_RAD.toarray(), columns=scaler.get_feature_names_out(['RAD']))
X_train_transform = X_train_transform.join(X_train_RAD_df)
X_test_RAD = scaler.transform(X_test[['RAD']])
X_test_RAD_df = pd.DataFrame(X_test_RAD.toarray(), columns=scaler.get_feature_names_out(['RAD']))
X_test_transform = X_test_transform.join(X_test_RAD_df)

Hay que notar que hemos creado dos dataframes distintos, uno para los datos de entrenamiento (train) y otro para los de prueba (test). Por otro lado, la variable dependiente, `MEDV`, la mantenemos separada, no tendría sentido incluirla como predictora de sí misma y les quitaría peso relativo al resto de variables, en tanto la que mejor explicaría `MEDV` sería `MEDV`, claro.

Finalmente, ahora ya tenemos todas nuestras variables independientes en dos conjuntos: X_train_transform y X_test_transform. Mientras que la variable dependiente la tenemos en y_train.MEDV e y_test.MEDV, con esto ya podemos entrenar nuestros modelos.

# *EX01* Crea almenys dos models de regressió diferents per intentar predir el millor possible el preu de les vivendes (MEDV) de l'arxiu adjunt.

In [20]:
import numpy as np
from sklearn.linear_model import LinearRegression

Utilizaremos dos tipos de regresión lineal múltiple, una de primer grado: modelo 1 (model) y otra polinómica de segundo grado: modelo 2 (model2).

# Modelo 1 - regresión lineal múltiple

En el primer modelo todas las variables del dataframe actúan como variables independientes (x) excepto MEDV, que actúa como variable dependiente (y).

In [29]:
x = np.array(X_train_transform)
y = np.array(y_train)

print(x.shape)
y.shape

(404, 21)


(404,)

Cargamos los datos para x e y, convirtiéndolos desde un dataframe (formato original) a dos arrays numpy, tal y como requiere el modelo. X es multidimensional (con 404 filas y 21 columnas), mientras y es unidimensional (con 404 filas).

In [31]:
model = LinearRegression().fit(x, y)

Ajustamos el modelo a los datos de x e y.

In [32]:
print(f"intercept: {model.intercept_}")
print(f"coefficients: {model.coef_}")

intercept: 26.594280052113863
coefficients: [ 0.25932839 -9.63770826  0.68346334 -4.03744671 -0.40771031  1.24641913
  2.20923065 -4.70053874 -3.11435146  0.1496336  -5.50543247  2.08733149
 -3.25897562 -1.25156236  1.29106492 -0.54594855 -0.37352386 -2.51365337
  1.75676861  0.36917488  4.52665534]


Obtenemos el punto de intersección y los coeficientes, que correponden a la estructura: y = b0 + b1·x1 + b2·x2 + ...+ b21·x21. En este caso: y = 26.59 + 0.25·CRIM - 9.637·ZN + ... + 4.52·RAD_24

# Modelo 2 - regresión múltiple polinomial de grado 2

El segundo modelo es una regresión múltiple polinomial de grado 2.

In [42]:
from sklearn.preprocessing import PolynomialFeatures

In [43]:
x_ = PolynomialFeatures(degree=2, include_bias=False).fit_transform(x)
x_

array([[0.23643695, 0.12962963, 0.03810505, ..., 0.        , 0.        ,
        0.        ],
       [0.17118768, 0.1399177 , 0.27085479, ..., 1.        , 0.        ,
        0.        ],
       [0.64662757, 0.64814815, 0.98043254, ..., 0.        , 0.        ,
        1.        ],
       ...,
       [0.64662757, 0.63374486, 0.95880536, ..., 0.        , 0.        ,
        1.        ],
       [0.21041056, 0.25102881, 0.60350154, ..., 1.        , 0.        ,
        0.        ],
       [0.2016129 , 0.2345679 , 0.39752832, ..., 0.        , 0.        ,
        0.        ]])

Transformamos la variable x para incluir los valores de x al cuadrado, y sin incluir el término de sesgo (include_bias=False), ya que por defecto ya lo aplica la regresión lineal que ajustamos a continuación.

In [44]:
model2 = LinearRegression().fit(x_, y)

Ajustamos el modelo ...

In [62]:
print(f"intercept: {model2.intercept_}")
print(f"coefficients: {model2.coef_}")

intercept: 27884373786.446785
coefficients: [-3.14631737e+11  3.35964161e+09 -5.25293350e+00  3.47520395e+10
  3.67476463e+01 -4.86231049e+09  2.98380280e+00 -1.85393829e+01
 -1.00933563e+09  9.83657837e-01  2.24794769e+00  1.67074203e+00
 -1.39421869e+10 -1.39421869e+10 -1.39421869e+10 -1.39421869e+10
 -1.39421869e+10 -1.33865867e+10  2.82639927e+10  1.26245597e+10
  4.22089687e+10  2.89235001e+01  6.72721939e+01  1.35045853e+01
  1.52367172e+01  4.43116455e+01 -2.84690094e+00  8.57063293e+00
  1.35340042e+01 -1.77408600e+00 -2.03761292e+00 -2.57736206e+00
 -1.49931183e+01  3.14631737e+11  3.14631737e+11  3.14631737e+11
  3.14631737e+11  3.14631737e+11  8.21795440e+10  3.60807678e+10
  2.62873020e+09  2.72934827e+10  2.50022984e+01 -3.05710335e+01
 -4.67205553e+01 -1.10451317e+00  9.94004822e+00  5.24969101e-02
  4.58684731e+01  2.46047211e+00 -8.61740112e-03  7.38985252e+00
 -1.00586858e+01 -3.35964161e+09 -3.35964166e+09 -3.35964159e+09
 -3.35964164e+09 -3.35964161e+09  4.25030624e+

... y de manera similar al modelo de grado 1, obtenemos los valores de intersección y coeficientes, que incluyen los valores de x, de x2 y de todas las intersecciones entre las x, de dos en dos: x1·x2, x1·x3, x1·x4 ... x20·x21. Por lo tanto, sería: 27884373786.446785 + -3.14631737e+11 · CRIM + ... + 4.22089687e+10 · RAD_23 · RAD_24

# *EX02* Compara’ls en base al MSE i al R2.

R2 en el primer modelo y segundo modelo, respectivamente:



In [80]:
print('R2 (regresión lineal): '+str(model.score(x, y).round(2)))
print('R2 (regresión polinomial): '+str(model2.score(x_, y).round(2)))

R2 (regresión lineal): 0.74
R2 (regresión polinomial): 0.95


Una R2 más próxima a 1 se interpreta en general como una mejor aproximación, lo cual haría al segundo modelo (polinomial) mejor que el primero en la previsión del precio de la vivienda. Por otro lado hay que tener en cuenta que también se corre el riesgo de sobreajuste del modelo a los datos, lo que podría llegar a suceder en el segundo caso con un valor de R2 de 0.94, pero esto lo comprobaremos en el ejercicio 4 con los datos de prueba que separamos (test).

MSE en el primer y segundo modelos respectivamente:

In [81]:
from sklearn.metrics import mean_squared_error
import math

y_ = model.predict(x)
y_2 = model2.predict(x_)

print('MSE (regresión lineal): '+str(mean_squared_error(y, y_).round(2)))
print('MSE (regresión polinomial): '+str(mean_squared_error(y, y_2).round(2)))

MSE (regresión lineal): 21.04
MSE (regresión polinomial): 4.25


Que el MSE sea de 21 significa que en promedio el modelo se equivoca en 4.58 miles de dólares (raíz de 21) en la previsión del precio de la vivienda en base a los valores de las variables.  
En este caso vemos como el error es significativamente mayor que el que se comete con un modelo cuadrático, de 2.06 miles de dólares (raíz de 4.24).
Es decir, 4.580 frente a 2.060 dólares de error en promedio. Mejor el modelo de grado 2 también según este segundo parámetro de medida del rendimiento.

# EX03 Entrena’ls utilitzant els diferents paràmetres que admeten per intentar millorar-ne la predicció.

# Modelo lineal

En primer lugar vamos a obtener los parámetros previos de nuestro modelo lineal:

In [69]:
print('Modelo lineal: '+str(model.get_params()))

Modelo lineal: {'copy_X': True, 'fit_intercept': True, 'n_jobs': None, 'positive': False}


Hay solamente dos parámetros que pueden afectar al rendimiento del modelo lineal: `fit_intercept` y `positive`, que se refieren a si queremos utilizar una constante en el modelo o no, y si los coeficientes del modelo deben ser necesariamente positivos o no. Vamos utilizar GridSearchCV para evaluarlos.

In [70]:
from sklearn.model_selection import GridSearchCV

In [72]:
param_grid = {
    'fit_intercept': [True, False],
    'positive': [True, False]
}
model0 = LinearRegression()
model_grid = GridSearchCV(model0, param_grid)

model_grid.fit(x, y)

print("Mejores parámetros:", model_grid.best_params_)

Mejores parámetros: {'fit_intercept': False, 'positive': False}


De acuerdo con el método GridSearchCV de la librería scikit-learn (sklearn), ambos parámetros deben establecerse como `False`, para optimizar el rendimiento del modelo lineal.

Entonces, vamos a obtener el nuevo rendimiento utilizando los parámetros óptimos:

In [75]:
print('Regresión lineal con parámetros óptimos:\n')
print('R2: '+str(model_grid.score(x, y).round(2)))
y_ = model_grid.predict(x)
print('MSE: '+str(mean_squared_error(y, y_).round(2)))

Regresión lineal con parámetros óptimos:

R2: 0.74
MSE: 21.04


Los rendimientos anteriores, sin cambiar los parámetros, eran exactamente los mismos para el modelo lineal, por lo tanto su variación no afecta a la capacidad predictiva del modelo lineal.

Seguimos con el modelo cuadrático:

In [77]:
print('Modelo polinomial: '+str(model2.get_params()))

Modelo polinomial: {'copy_X': True, 'fit_intercept': True, 'n_jobs': None, 'positive': False}


Arriba los parámetros por defecto. Abajo los parámetros óptimos.

In [78]:
model_grid_2 = GridSearchCV(model0, param_grid)
model_grid_2.fit(x_, y)
print("Mejores parámetros:", model_grid_2.best_params_)

Mejores parámetros: {'fit_intercept': False, 'positive': True}


A continuación medimos el rendimiento del modelo con los nuevos parámetros:

In [79]:
print('Regresión polinomial con parámetros óptimos:\n')
print('R2: '+str(model_grid_2.score(x_, y).round(2)))
y_2_grid = model_grid_2.predict(x_)
print('MSE: '+str(mean_squared_error(y, y_2_grid).round(2)))

Regresión polinomial con parámetros óptimos:

R2: 0.8
MSE: 16.04


Recordemos que el rendimiento para el modelo polinomial con los parámetros por defecto era: R2=0.95 y MSE=4.25

Por lo tanto, no sólo no mejoramos el rendimiento cambiando el valor de los parámetros, sino que lo empeoramos, siendo la explicación más inmediata la de un sobreajuste del modelo a los datos de entrenamiento que se da al optimizarlos.

# EX04. Compara el seu rendiment emprant l’aproximació train/test o emprant totes les dades (validació interna).

# **Modelo lineal**

En este caso vamos a hacer la comparación sin transformar los datos, es decir, con los datos originales.

In [105]:
print('Rendimiento del modelo lineal utilizando todos los datos:\n')
modelo_completo = LinearRegression().fit(X, viviendas.MEDV)
print('R2: '+str(modelo_completo.score(X, viviendas.MEDV)))
y_prevista = modelo_completo.predict(X)
print('MSE: '+str(mean_squared_error(viviendas.MEDV, y_prevista)))

Rendimiento del modelo lineal utilizando todos los datos:

R2: 0.7406426641094095
MSE: 21.894831181729202


In [107]:
print('Rendimiento del modelo lineal utilizando train/test:\n')

modelo_train_test = LinearRegression().fit(X_train, y_train)
print('R2: '+str(modelo_train_test.score(X_test, y_test)))
y_train_test = modelo_train_test.predict(X_test)
print('MSE: '+str(mean_squared_error(y_test, y_train_test)))

Rendimiento del modelo lineal utilizando train/test:

R2: 0.7634174432138461
MSE: 23.380836480270336


Podemos ver como el rendimiento del modelo lineal es algo inferior utilizando el modelo train/test que utilizando todos los datos, lo que tiene su sentido al trabajar con un número mayor de observaciones para hacer las predicciones.

# Modelo polinomial:

In [108]:
print('Rendimiento del modelo polinomial utilizando todos los datos:\n')
X_completo = PolynomialFeatures(degree=2, include_bias=False).fit_transform(X)
modelo_poli_completo = LinearRegression().fit(X_completo, viviendas.MEDV)
print('R2: '+str(modelo_poli_completo.score(X_completo, viviendas.MEDV)))
y_poli_completo = modelo_poli_completo.predict(X_completo)
print('MSE: '+str(mean_squared_error(viviendas.MEDV, y_poli_completo)))

Rendimiento del modelo polinomial utilizando todos los datos:

R2: 0.9289961714593018
MSE: 5.994111690794227


In [109]:
print('Rendimiento del modelo polinomial utilizando train/test:\n')
X_poli_train = PolynomialFeatures(degree=2, include_bias=False).fit_transform(X_train)
modelo_poli_train_test = LinearRegression().fit(X_poli_train, y_train)
X_poli_test = PolynomialFeatures(degree=2, include_bias=False).fit_transform(X_test)
print('R2: '+str(modelo_poli_train_test.score(X_poli_test, y_test)))
y_poli_test = modelo_poli_train_test.predict(X_poli_test)
print('MSE: '+str(mean_squared_error(y_test, y_poli_test)))

Rendimiento del modelo polinomial utilizando train/test:

R2: 0.9116244055454207
MSE: 8.733929292415759


El modelo polinomial también muestra un rendimiento algo superior utilizando todos los datos que con train/test.

# EX05 No facis servir la variable del nombre d'habitacions (RM) a l’hora de fer prediccions.

Vamos a utilizar el modelo polinomial con train/test, ya que sería el que generalmente utilizaríamos, para evaluar las diferencias:

En primer lugar redefinimos X:

In [112]:
X_train_sin_RM = X_train.drop('RM', axis=1)
X_test_sin_RM = X_test.drop('RM', axis=1)

In [113]:
print('Modelo polinomial utilizando train/test sin `RM`:\n')
X_poli_train_sin_RM = PolynomialFeatures(degree=2, include_bias=False).fit_transform(X_train_sin_RM)
modelo_poli_sin_RM = LinearRegression().fit(X_poli_train_sin_RM, y_train)
X_poli_test_sin_RM = PolynomialFeatures(degree=2, include_bias=False).fit_transform(X_test_sin_RM)
print('R2: '+str(modelo_poli_sin_RM.score(X_poli_test_sin_RM, y_test)))
y_poli_sin_RM = modelo_poli_sin_RM.predict(X_poli_test_sin_RM)
print('MSE: '+str(mean_squared_error(y_test, y_poli_sin_RM)))

Modelo polinomial utilizando train/test sin `RM`:

R2: 0.8262413537202487
MSE: 17.172113408902707


Vemos como el rendimiento disminuye en el caso de no utilizar `RM`, lo que refleja su alto valor predictivo del precio.

# Síntesis

| Modelo de Regresión | R2 | MSE |
| ------------------- | -- | --- |
| Regresión Lineal con datos transformados y parámetros por defecto | 0.74 | 21.04 |
| Regresión Polinomial con datos transformados y parámetros por defecto | 0.95 | 4.25 |
| Regresión Lineal con parámetros óptimos y datos transformados | 0.74 | 21.04 |
| Regresión Polinomial con parámetros óptimos, modelo train/test y datos transformados | 0.8 | 16.04 |
| Modelo Lineal utilizando todos los datos pero sin transformarlos | 0.74 | 21.89 |
| Modelo Lineal utilizando train/test sin transformar los datos | 0.76 | 23.38 |
| Modelo Polinómico utilizando todos los datos | 0.92 | 5.99 |
| Modelo Polinómico utilizando train/test | 0.91 | 8.73 |
| Modelo Polinómico utilizando train/test sin `RM` | 0.82 | 17.17 |


Resumiendo ...

Como podemos comprobar en la tabla anterior, que sintetiza las variaciones realizadas, el modelo que mejor puede predecir el precio de las viviendas en los suburbios de Boston en los años 70 de entre los utilizados, es el polinómico de grado dos con los datos transformados (0.95 y 4.25), al que le seguiría el mismo modelo con las variables sin transformar pero utilizando todos los datos (0.92 y 5.99), lo que no es recomendable, pues se puede incurrir en un sobreajuste, y no representa una gran mejora respecto al modelo train/test polinomial con los datos sin transformar (0.91 y 8.73). Los modelos lineales presentan un rendimiento claramente inferior y optimizar los parámetros en este caso empeora el rendimiento.